baseapp/abci_test.go

```diff
@@ -737,6 +737,121 @@ func TestABCI_FinalizeBlock_MultiMsg(t *testing.T) {
 	require.Equal(t, int64(2), msgCounter2)
 }
 
+func TestABCI_DeliverTx_NonAtomicMultiMsg(t *testing.T) {
+	anteKey := []byte("ante-key")
+	anteOpt := func(bapp *baseapp.BaseApp) { bapp.SetAnteHandler(anteHandlerTxTest(t, capKey1, anteKey)) }
+	suite := NewBaseAppSuite(t, anteOpt)
+
+	_, err := suite.baseApp.InitChain(&abci.RequestInitChain{
+		ConsensusParams: &cmtproto.ConsensusParams{},
+	})
+	require.NoError(t, err)
+
+	deliverKey := []byte("deliver-key")
+	baseapptestutil.RegisterCounterServer(suite.baseApp.MsgServiceRouter(), CounterServerImpl{t, capKey1, deliverKey})
+
+	// run a multi-msg tx
+	// with all msgs the same route
+	tx := newTxCounter(t, suite.txConfig, NonAtomic, 0, 0, 1, 2)
+	txBytes, err := suite.txConfig.TxEncoder()(tx)
+	require.NoError(t, err)
+
+	res, err := suite.baseApp.FinalizeBlock(&abci.RequestFinalizeBlock{
+		Height: 1,
+		Txs:    [][]byte{txBytes},
+	})
+	require.NoError(t, err)
+
+	require.True(t, res.TxResults[0].IsOK(), fmt.Sprintf("%v", res))
+
+	store := getFinalizeBlockStateCtx(suite.baseApp).KVStore(capKey1)
+
+	// tx counter only incremented once
+	txCounter := getIntFromStore(t, store, anteKey)
+	require.Equal(t, int64(1), txCounter)
+
+	// msg counter incremented three times
+	msgCounter := getIntFromStore(t, store, deliverKey)
+	require.Equal(t, int64(3), msgCounter)
+
+	_, _, addr := testdata.KeyTestPubAddr()
+
+	testCases := []struct {
+		name         string
+		succeed      []bool
+		finalCounter int64
+		expErr       bool
+	}{
+		{"all messages succeed", []bool{true, true}, 5, false},
+		{"all messages succeed2", []bool{true, true}, 7, false},
+		{"first succeed", []bool{true, false}, 8, false},
+		{"second succeed", []bool{false, true}, 9, false},
+		{"all messages fail", []bool{false, false}, 9, true},
+	}
+
+	for testNum, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
```

Comment:
Please make a fesh copy of tc and testNum that ensure that the values will be afresh even if the test is parallelized, so:
```go
   testNum, tc := testNum, tc
   t.Run(tc.name, func(t *testing.T) {
       ...
```